### DISCUSSION TITLE
VerneMQ high memory usage

### DISCUSSION THREAD

**Discussion by vshirol** [vernemq/vernemq#2034] [Q&A]
I am trying to understand memory usage by VerneMQ. Below are the more details:

Environment
VerneMQ 1.10.2, single node bare metal 
OS: RHEL 7.8
Server: 
   CPU(s): 24  
   Thread(s) per core: 2  
   Core(s) per socket: 6 
   Socket(s): 2 
   Memory 64GB

Test:
1+ million clients connection creation using benchmark tool (running from different machine with multiple IP interface)

Result:
VMQ max connections achieved: 1 million
Connect Rate achieved : 900/sec
VMQ System CPU usage : 18%
VMQ Memory usage : 43GB
VMQ bandwidth usage :  12mbps
Load Balancer : No. As we are testing for single node capacity

We felt VMQ memory usage looks very high, although in this test its only MQTT client's connecting to broker (successfully) and no client is subscribing to any topic. That means there is no pub/sub involved in this test.

Below are the TCP/OS tuning done before tests:
sysctl.conf:
net.ipv4.tcp_rmem= 4096 16384 32768
net.ipv4.tcp_wmem= 4096 16384 32768

Also please find attached vernemq.conf (zip) file. Let me know any more details needed and let me know if any configuration changes required as well.
[vernemq.zip](https://github.com/vernemq/vernemq/files/9533071/vernemq.zip)

**Comment by ioolkos** [comment]
@vshirol Thanks. No surprise in 1M connections taking a lot of RAM. You'll have a rcvbuf a sndbuf and an internal buffer per TCP connection.
If this is 16 KB per buffer, you'll have 48 KB * 1M. 

If you want to control buffers directly (to possibly lower the sizes), you can use the `buffer_sizes` setting in the vernemq.conf file to set defaults for a listener.
I have no bandwidth to review your logs/config files but I think above note helps you provision and project memory closely enough.

**Reply by vshirol** [reply]
Thanks @ioolkos. I will look into buffer_sizes setting in the vernemq.conf file.
